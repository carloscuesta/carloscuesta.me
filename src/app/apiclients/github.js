'use strict';

const ApiClient = require('./apiclient')
const CacheApiClient = require('./cache')
const nodeCache = require('memory-cache')

const CACHE_TIME = 86400000
const GITHUB_CACHE = 'GITHUB_CACHE'
const githubDataCache = nodeCache.get(GITHUB_CACHE)

class GithubApiClient {
	constructor() {
		this.apiClient = new ApiClient({
			'base_url': 'https://api.github.com',
			'cache': 5
		});
	}

	getSearch(params) {
		return CacheApiClient.validate
			.call(this.apiClient, '/search/repositories', params)
	}

	mutator(payload) {
		if (githubDataCache) {
			return githubDataCache
		}

		const repositories = payload.items.map((repo) => ({
			language: repo.language && repo.language.toLowerCase(),
			url: repo.html_url,
			name: repo.name,
			stars: repo.stargazers_count,
			forks: repo.forks,
			description: repo.description,
			homepage: repo.homepage
		}))

		nodeCache.put(GITHUB_CACHE, repositories, CACHE_TIME)
		return repositories
	}
}

module.exports = new GithubApiClient()
