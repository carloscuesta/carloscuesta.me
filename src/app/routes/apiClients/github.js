const nodeCache = require('memory-cache')
const ApiClient = require('./apiClient')
const CacheApiClient = require('./cache')
const CONFIG = require('./config')
const githubDataCache = nodeCache.get(CONFIG.GITHUB_CACHE)

class GithubApiClient {
  constructor () {
    this.apiClient = new ApiClient({ 'base_url': 'https://api.github.com' })
  }

  getSearch (params) {
    return CacheApiClient.validate.call(
      this.apiClient, '/search/repositories',
      params
    )
  }

  mutator (payload) {
    if (githubDataCache) return githubDataCache

    const repositories = payload.items.map((repo) => ({
      language: repo.language && repo.language.toLowerCase(),
      url: repo.html_url,
      name: repo.name,
      stars: repo.stargazers_count,
      forks: repo.forks,
      description: repo.description,
      homepage: repo.homepage
    }))

    nodeCache.put(CONFIG.GITHUB_CACHE, repositories, CONFIG.CACHE_TIME)
    return repositories
  }
}

module.exports = new GithubApiClient()
