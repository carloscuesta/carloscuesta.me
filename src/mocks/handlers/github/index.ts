import { rest } from 'msw'

import repositories from './mocks/repos.json'
import user from './mocks/user.json'

const baseUrl = 'https://api.github.com'

const githubHandlers = [
  rest.get(`${baseUrl}/users/carloscuesta/repos`, (_req, res, ctx) =>
    res(ctx.json(repositories))
  ),
  rest.get(`${baseUrl}/users/carloscuesta`, (_req, res, ctx) =>
    res(ctx.json(user))
  ),
]

export default githubHandlers
