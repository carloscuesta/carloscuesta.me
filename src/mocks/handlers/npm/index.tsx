import { rest } from 'msw'

import packages from './mocks/packages.json'
import downloads from './mocks/downloads.json'

const npmHandlers = [
  rest.get('https://registry.npmjs.org/-/v1/search', (_req, res, ctx) =>
    res(ctx.json(packages)),
  ),
  rest.get(
    'https://api.npmjs.org/downloads/point/last-year/*',
    (_req, res, ctx) => res(ctx.json(downloads)),
  ),
]

export default npmHandlers
