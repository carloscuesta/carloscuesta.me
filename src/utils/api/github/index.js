// @flow
import callApi, { type callApiOptions } from 'src/utils/api/callApi'
import {
  transformRepositories,
  transformUserInformation,
  type Repository,
  type UserInformation
} from './mutators'

const githubApiClient = (options: callApiOptions): Promise<Object> => callApi({
  mutator: options.mutator,
  url: `https://api.github.com${options.url}`,
  requestOptions: {
    headers: { Accept: 'application/vnd.github.v3+json' }
  }
})

export const fetchRepositories = (): Promise<Array<Repository>> => githubApiClient({
  mutator: transformRepositories,
  url: '/users/carloscuesta/repos'
})

export const fetchUserInformation = (): Promise<UserInformation> => githubApiClient({
  mutator: transformUserInformation,
  url: '/users/carloscuesta'
})
