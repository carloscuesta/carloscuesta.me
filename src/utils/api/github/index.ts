import callApi, { type callApiOptions } from 'src/utils/api/callApi'
import { transformRepositories, transformUserInformation } from './mutators'

const githubApiClient = <Response, Data>(
  options: callApiOptions<Response, Data>,
) =>
  callApi({
    mutator: options.mutator,
    url: `https://api.github.com${options.url}`,
    requestOptions: {
      headers: { Accept: 'application/vnd.github.v3+json' },
    },
  })

export const fetchRepositories = () =>
  githubApiClient({
    mutator: transformRepositories,
    url: '/users/carloscuesta/repos',
  })

export const fetchUserInformation = () =>
  githubApiClient({
    mutator: transformUserInformation,
    url: '/users/carloscuesta',
  })
