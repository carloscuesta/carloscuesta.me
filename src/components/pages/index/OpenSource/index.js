// @flow
import React, { type Element } from 'react'

import { type Repository } from 'src/utils/api/github/mutators'
import Wrapper from 'src/components/shared/Wrapper'
import SectionTitle from 'src/components/shared/SectionTitle'
import Project from './Project'

type Props = {
  repositories: Array<Repository>
}

const OpenSource = (props: Props): Element<'section'> => (
  <section>
    <Wrapper>
      <SectionTitle
        title='Open Source'
        subTitle='Featured open source projects.'
      />

      <div className='row'>
        {props.repositories.map((repository) => (
          <Project repository={repository} key={repository.name} />
        ))}
      </div>
    </Wrapper>
  </section>
)

export default OpenSource
