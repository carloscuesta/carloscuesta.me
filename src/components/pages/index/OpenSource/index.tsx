import { type Repository } from 'src/utils/api/github/mutators'
import Wrapper from 'src/components/shared/Wrapper'
import SectionTitle from 'src/components/shared/SectionTitle'
import Project from './Project'

type Props = {
  repositories: Repository[]
}

const OpenSource = (props: Props) => (
  <section>
    <Wrapper>
      <SectionTitle
        title='Open Source'
        subTitle='Featured open source projects.'
        viewAllLink='/opensource'
      />

      <div className='row'>
        {props.repositories.map((repository) => (
          <Project repository={repository} key={repository.name} />
        ))}
      </div>
    </Wrapper>
  </section>
)

export default OpenSource
