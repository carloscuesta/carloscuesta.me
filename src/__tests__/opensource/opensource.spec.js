import renderer from 'react-test-renderer'

import OpenSource, { getStaticProps } from 'src/pages/opensource'
import { fetchRepositories, fetchUserInformation } from 'src/utils/api/github'
import { fetchPublishedPackages, fetchDownloadsCount } from 'src/utils/api/npm'
import * as stubs from './stubs'

jest.mock('next-seo', () => ({ NextSeo: 'NextSeo' }))
jest.mock('src/utils/api/blog')
jest.mock('src/utils/api/github')
jest.mock('src/utils/api/npm')

/*
  I'm mocking the components since those are unit tested.
  So there's no point to test the rendering of them twice.
  It's more easier to manage little snapshots than big ones.
*/

jest.mock('src/components/shared/PageTitle', () => 'PageTitle')
jest.mock('src/components/pages/opensource/Repositories', () => 'Repositories')
jest.mock('src/components/pages/opensource/Stats', () => 'Stats')

describe('index', () => {
  describe('page', () => {
    it('should match opensource page', () => {
      const wrapper = renderer.create(<OpenSource {...stubs.props} />)

      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('getStaticProps', () => {
    beforeAll(() => {
      fetchRepositories.mockReturnValue(stubs.repositories)
      fetchUserInformation.mockReturnValue(stubs.userInformation)
      fetchPublishedPackages.mockReturnValue(stubs.publishedPackages)
      fetchDownloadsCount.mockReturnValue(stubs.packageDownloads)
    })

    it('should return repositories and userInformation as props and set revalidate', async () => {
      const props = await getStaticProps()

      expect(fetchRepositories).toHaveBeenCalledWith()
      expect(fetchUserInformation).toHaveBeenCalledWith()
      expect(fetchPublishedPackages).toHaveBeenCalledWith()
      expect(fetchDownloadsCount).toHaveBeenCalledWith(stubs.publishedPackages)

      expect(props).toEqual({
        props: {
          repositories: stubs.repositories,
          userInformation: stubs.userInformation,
          packageDownloads: stubs.packageDownloads
        },
        revalidate: 3600
      })
    })
  })
})
