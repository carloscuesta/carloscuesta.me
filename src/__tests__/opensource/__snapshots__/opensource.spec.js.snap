// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index page should match opensource page 1`] = `
Array [
  <NextSeo
    canonical="https://carloscuesta.me/opensource"
    title="Carlos Cuesta – Open Source"
  />,
  <main>
    <div
      className="wrapper"
    >
      <PageTitle
        title="Open Source"
      />
      <Stats
        followers={100}
        packageDownloads={10000}
        stars={10624}
      />
      <Repositories
        repositories={
          Array [
            Object {
              "description": "An emoji guide for your commit messages. 😜 ",
              "language": "javascript",
              "name": "gitmoji",
              "stars": 7460,
              "url": "https://github.com/carloscuesta/gitmoji",
            },
            Object {
              "description": "A gitmoji interactive command line tool for using emojis on commits. 💻",
              "language": "javascript",
              "name": "gitmoji-cli",
              "stars": 2254,
              "url": "https://github.com/carloscuesta/gitmoji-cli",
            },
            Object {
              "description": "A material design theme for your terminal. ✨",
              "language": "shell",
              "name": "materialshell",
              "stars": 710,
              "url": "https://github.com/carloscuesta/materialshell",
            },
            Object {
              "description": "A Front End development Gulp.js based workflow. 🚀",
              "language": "javascript",
              "name": "starterkit",
              "stars": 84,
              "url": "https://github.com/carloscuesta/starterkit",
            },
            Object {
              "description": "A material design theme for Hyper based on materialshell. ✨",
              "language": "javascript",
              "name": "hyper-materialshell",
              "stars": 69,
              "url": "https://github.com/carloscuesta/hyper-materialshell",
            },
            Object {
              "description": "A simple and reusable React-Native error boundary component 🐛",
              "language": "javascript",
              "name": "react-native-error-boundary",
              "stars": 47,
              "url": "https://github.com/carloscuesta/react-native-error-boundary",
            },
          ]
        }
      />
    </div>
  </main>,
]
`;
